File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.js
/**
 * @param {string[]} tokens
 * @return {number}
 */

// an array string
// int result
// stack = last in first out
// push into stack until we an operation ( + - * /)
// operation pop until 2 numbers to use for that operation
// operation I have mindful of the order of num being popped
// + : get two number add it and push it back
// * : get two number multiple and push it back 
// / : second num popped / first num popped and push it back rounding down (floor)
// - : second num popped - first num popped and push it back
// 
// when we empty stakc return the result

// T of N 
// S of N(stack)
// 
var evalRPN = function(tokens) {
    let stack = []
    
    for (let token of tokens){
        console.log(stack)
        if(token==="+"){//+
            let num1= stack.pop()
            let num2= stack.pop()
            let sum = num1 + num2
            stack.push(sum)
        } else if(token==="-"){
            let num1= stack.pop()
            let num2= stack.pop()
            let dif = num2 - num1
            stack.push(dif)
        } else if(token==="*"){
            let num1= stack.pop()
            let num2= stack.pop()
            let pro = num2 * num1
            stack.push(pro)
        } else if(token==="/"){
            let num1= stack.pop()
            let num2= stack.pop()
            let div = Math.trunc(num2 / num1)
            stack.push(div)
        } else {
            let int = Number(token)
            stack.push(int)
        }
    }
    return stack[0]
};

// []
// [ 10 ]
// [ 10, 6 ]
// [ 10, 6, 9 ]
// [ 10, 6, 9, 3 ] +
// [ 10, 6, 12 ]
// [ 10, 6, 12, -11 ] *
// [ 10, 6, -132 ]
// [ 10, -1 ]
// [ -10 ]
// [ -10, 17 ]
// [ 7 ]
// [ 7, 5 ]

File: 0150-evaluate-reverse-polish-notation/0150-evaluate-reverse-polish-notation.py
# stack = result, 3
# + operation result = stack[-1] + stack[-1] 
# * operation = result * 3 = 
# put it back
# round down
# stack to keep track of past values and result
# when we hit any kind of operation, we going access last two position and calc according
# push the result back into stack
# assuming 1 result in the stack return that 
# performance = N
# space = N 
# 
# 

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []

        for token in tokens:
            if token == "+":
                second = stack.pop()
                first = stack.pop()
                result = first + second
            elif token == "-":
                second = stack.pop()
                first = stack.pop()
                result = first - second
            elif token == "*":
                second = stack.pop()
                first = stack.pop()
                result = first * second
            elif token == "/":
                second = stack.pop()
                first = stack.pop()
                result = first / second
                if result > 0:
                    result = math.floor(result)
                else:
                    result = math.ceil(result)        
            else: 
                result = int(token)
            stack.append(result)
            # print(stack)

        return stack[0]

        # tokens =
# ["10","6","9","3","+","-11","*","/","*","17","+","5","+"]

# Use Testcase
# Stdout
# [10]
# [10, 6]
# [10, 6, 9]
# [10, 6, 9, 3]
# [10, 6, 12] + 
# [10, 6, 12, -11] *
# [10, 6, -132]  
# [10, -1]
# [-10]
# [-10, 17]
# [7]
# [7, 5]
# [12
        